rfd/0025-hsm.md

```diff
@@ -279,36 +280,61 @@ API resource as read-only.
-the CA rotation, Auth servers will not use the HSM for signing even when
-[configured](#configuration). CA rotation is needed to propagate the CA
-certificates associated with the new private key to all users and hosts.
-Without the rotation, all certificates issued by this Auth server would not be
-trusted by any client.
+Migration of existing clusters to use HSMs will involve a CA rotation. CA
+rotation is needed to propagate the CA certificates associated with the new
+private key to all users and hosts.  Without the rotation, all certificates
+issued by this Auth server will not be trusted by any client.
 
 On startup, Auth server will fetch existing CAs from the backend and:
 
 - if no CAs exist, create a key in HSM and register it in `ActiveKeys`
   - this takes care of brand new Teleport clusters
 - if CAs exist, but _don't_ have an HSM key from this host registered in
-  `ActiveKeys`, require a CA rotation to add the key
-  - print a warning on startup and in `tctl status` until HSM key is added
-  - all signing operations against this Auth server will fail until rotation;
-    when adding a new HSM-enabled Auth server to the cluster, admins should not
-    route any traffic to it until the rotation is complete
-  - for HA setup in a new cluster, admins can run `tctl auth rotate
-    --grace-period=0` to perform fast rotation after adding all Auth servers
+  `ActiveKeys`, create a key in HSM and register it in `AdditionalTrustedKeys`
+  - this key will be used to sign the Admin identity, necessary for tctl
+  - this key is not expected to be trusted by the rest of the cluster until a CA
+    rotation is completed, but will be trusted by the local auth server.
+  - print a warning on startup and in `tctl status` until a CA rotation is
+    completed.
+  - all user and jwt signing operations against this Auth server will fail until
+    rotation; when adding a new HSM-enabled Auth server to the cluster, admins
+    should not route any traffic to it until the rotation is complete
 - if CAs exist and have an HSM key from this host, load the key and use it for
   signing
   - if the key is not found in the HSM (e.g. it was factory reset), the Auth
-    server will exit; to resolve this, admins have remove
+    server will exit; to resolve this, admins have to remove
     `/var/lib/teleport/host_uuid` to explicitly change the UUID and
-    dis-associate the old HSM key from this instance; from there, Auth server
+	dis-associate the old HSM key from this instance; from there the Auth server
     behaves as above when registering a new key
 
 When rotation is initiated through a different Auth server, Auth servers with
 HSMs will add a new PKCS#11 key reference to CA `AdditionalTrustedKeys` in the
-`init` phase of rotation.
+`init` phase of rotation. Admins should not perform a rotation with
+`--grace_period=0` (or any very short value, less than 1 minute) so that all auth
```

Comment:
we should make this explicitly fail and require a non-zero grace period

Reply:
Maybe I'm doing something wrong but I just realized we don't even allow a grace period of 0 at all, looks like we unconditionally hit this https://github.com/gravitational/teleport/blob/9b8b9d6d0c115d43d31d53c47db3050e27edbc4a/api/types/authority.go#L563-L565

Reply:
huh, looks like our CLI help text is out of date then